### PR TITLE
Limit to single worker pool per version of babel-core

### PR DESCRIPTION
**Pull request by mikrostew**
Currently there is one worker pool per installation of broccoli-babel-transpiler. For projects where all dependencies are on the same babel version this is not a problem.

However, when dependencies require a mix of babel 5.x and 6.x (for example during upgrade), this can result in dozens of installations of broccoli-babel-transpiler. Since there is one pool per installation, this results in way too many workers.

This PR limits the number of worker pools to one per version of babel-core. The pool state is stored on the global 'process' as 'v1/broccoli-babel-transpiler/workerpool/babel-core-X.X.X'.

### CODE REVIEW

File: lib/parallel-api.js [hunk 1]
```diff
@@ -8,12 +9,31 @@ var debugGenerator = require('heimdalljs-logger');
...
-var pool = workerpool.pool(path.join(__dirname, 'worker.js'), { maxWorkers: jobs });
-
 var loggerName = 'broccoli-persistent-filter:ParallelApi';
 var _logger = debugGenerator(loggerName);
 
+var babelCoreVersion = getBabelVersion();
+
+// return the version of Babel that will be used by this plugin
+function getBabelVersion() {
+  var babelCoreRequirePath = require.resolve('babel-core');
```
Comment (stefanpenner) [inline lib/parallel-api.js]: I believe `require.require('babel-core/package.json').version;` is the correct enchantation to consistently gain access to the version.


`require.resolve('babel-core')` will resolve the packages `main` file, which may reside more nested

File: lib/parallel-api.js [hunk 2]
```diff
@@ -8,12 +9,31 @@ var debugGenerator = require('heimdalljs-logger');
...
+  var babelCorePackageJson = path.join(path.dirname(babelCoreRequirePath), 'package.json');
+  var packageJson = fs.readFileSync(babelCorePackageJson);
+  return JSON.parse(packageJson).version;
+}
+
+function getWorkerPool() {
```
Comment (stefanpenner) [inline lib/parallel-api.js]: LGTM

**Comment by stefanpenner** [comment]
rebuilding appveyor, but it may be a legit failure

File: test.js [hunk 3]
```diff
@@ -1230,3 +1231,70 @@ describe('concurrency', function() {
     expect(ParallelApi.jobs).to.equal(17);
   });
 });
+
+describe('getBabelVersion()', function() {
+  it ('returns the correct version', function() {
+    var babelCorePath = path.join(__dirname, 'node_modules', 'babel-core', 'index.js');
```
Comment (stefanpenner) [inline test.js]: I believe the following would be simpler:

```
var version = require(__dirname + 'node_modules/babel-core/package.json').version
```

File: test.js [hunk 4]
```diff
@@ -1230,3 +1231,70 @@ describe('concurrency', function() {
...
+      'transform-strict-mode',
+      'transform-es2015-block-scoping'
+    ]
+  };
+
+  before(function() {
+    // shut down any workerpool that is running at this point
+    var babelCoreVersion = ParallelApi.getBabelVersion();
+    var something = 'v1/broccoli-babel-transpiler/workerpool/babel-core-' + babelCoreVersion;
+    var runningPool = process[something];
```
Comment (stefanpenner) [inline test.js]: rather the cleaning up when the next test task starts, we should always cleanup when the current test task ends.
This ensures our tests leave themselves in a pristine state.

So specifically, the `terminate` and `delete` should be in `after` or `afterEach`

File: test.js [hunk 5]
```diff
@@ -1230,3 +1231,70 @@ describe('concurrency', function() {
...
+      ParallelApiTwo.transformString(stringToTransform, options),
+      ParallelApiTwo.transformString(stringToTransform, options),
+    ]).then(function() {
+      return new Promise(function(resolve, reject) {
+        // find all worker processes
+        ps.lookup({
+          command: 'node',
+          arguments: path.join('broccoli-babel-transpiler', 'lib', 'worker.js'),
+        }, function(err, resultList ) {
+          expect(err).to.eql(null);
```
Comment (stefanpenner) [inline test.js]: lets `denodeify` ps.lookup, that way it is nicely promisified. Which will both simplify and improve the safety of our test.

```
let lookup = RSVP.denodeify(ps.lookup);
```

File: lib/parallel-api.js (hunk 1 above)
Comment (mikrostew) [reply lib/parallel-api.js]: that worked :)

File: test.js (hunk 3 above)
Comment (mikrostew) [reply test.js]: yeah much simpler than what I had

File: test.js (hunk 4 above)
Comment (mikrostew) [reply test.js]: moved the workerpool cleanup to after() for the tests that use it

File: test.js (hunk 5 above)
Comment (mikrostew) [reply test.js]: ah good point, done

**Comment by mikrostew** [comment]
Not sure if the AppVeyor failure is a legit failure, or that ps-node isn't finding the processes on windows. Will get a working windows VM to test this

**Comment by mikrostew** [comment]
I was able to repro this on Windows. Test is failing because ps-node is using paths that look like 'C:\\User\\blah\\broccoli-babel-transpiler\\lib\\worker.js', with 2 backslashes for some reason

File: test.js [hunk 6]
```diff
@@ -1230,3 +1262,61 @@ describe('concurrency', function() {
...
+  };
+
+  after(function() {
+    return terminateWorkerPool();
+  });
+
+  it('should limit to one pool per babel version', function() {
+    this.timeout(5*1000);
+    delete require.cache[parallelApiPath];
+    process.env.JOBS = '2';
```
Comment (stefanpenner) [inline test.js]: in `afterEach` we should restore `JOBS` it what is was before this test rand.

File: test.js [hunk 7]
```diff
@@ -1230,3 +1262,61 @@ describe('concurrency', function() {
...
+  var options = {
+    inputSourceMap: false,
+    sourceMap: false,
+    plugins: [
+      'transform-strict-mode',
+      'transform-es2015-block-scoping'
+    ]
+  };
+
+  after(function() {
```
Comment (stefanpenner) [inline test.js]: should this be `afterEach`?

Comment (mikrostew) [reply test.js]: oops it should be

File: test.js (hunk 6 above)
Comment (mikrostew) [reply test.js]: good catch - will add that so these tests are not leaking state

**Comment by stefanpenner** [comment]
@mikrostew thanks for working on ^, this will help those upgrading and having both versions active ALOT.

**Pull request merged**